Modules/socketmodule.c

```diff
@@ -305,6 +305,69 @@ if_indextoname(index) -- return the corresponding interface name\n\
 /* Provides the IsWindows7SP1OrGreater() function */
 #include <VersionHelpers.h>
 
+/* This block removes some flags on older version Windows during run-time.
+   https://msdn.microsoft.com/en-us/library/windows/desktop/ms738596.aspx */
+#if 1
+typedef struct {
+    DWORD build_number;  /* available starting with this Win10 BuildNumber */
+    const char flag_name[20];
+} FlagRuntimeInfo;
+
+/* IMPORTANT: make sure the list ordered by descending build_number */
+static FlagRuntimeInfo flags[] = {
+    /* available starting with Windows 10 1709 */
+    {16299, "TCP_KEEPIDLE"},
+    {16299, "TCP_KEEPINTVL"},
+    /* available starting with Windows 10 1703 */
+    {15063, "TCP_KEEPCNT"},
+    /* available starting with Windows 10 1607 */
+    {14393, "TCP_FASTOPEN"}
+};
+
+static PyObject *
+remove_unusable_flags(PyObject *m)
+{
+    PyObject *dict;
+    OSVERSIONINFOEX info;
+    DWORDLONG dwlConditionMask;
+
+    dict = PyModule_GetDict(m);
+    if (dict == NULL) {
+        return m;
+    }
+
+    /* set to Windows 10, except BuildNumber. */
+    memset(&info, 0, sizeof(info));
+    info.dwOSVersionInfoSize = sizeof(info);
+    info.dwMajorVersion = 10;
+    info.dwMinorVersion = 0;
+
+    /* set Condition Mask */
+    dwlConditionMask = 0;
+    VER_SET_CONDITION(dwlConditionMask, VER_MAJORVERSION, VER_GREATER_EQUAL);
+    VER_SET_CONDITION(dwlConditionMask, VER_MINORVERSION, VER_GREATER_EQUAL);
+    VER_SET_CONDITION(dwlConditionMask, VER_BUILDNUMBER, VER_GREATER_EQUAL);
+
+    for (int i = 0; i < sizeof(flags)/sizeof(FlagRuntimeInfo); i++) {
+        info.dwBuildNumber = flags[i].build_number;
+        /* greater than or equal to the specified version? 
+           Compatibility Mode will not cheat VerifyVersionInfo(...) */
+        if (VerifyVersionInfo(
```

Comment:
The `IsWindowsVersionOrGreater` function should always be available, which will simplify this call.

Reply:
[`IsWindowsVersionOrGreater`](https://msdn.microsoft.com/en-us/library/windows/desktop/dn424964(v=vs.85).aspx) is only aware of `MajorVersion`, `MinorVersion`, `ServicePackMajor`. While we need to compare `BuildNumber`.

`IsWindowsVersionOrGreater` uses [`VerifyVersionInfo`](https://msdn.microsoft.com/en-us/ms725492(VS.85).aspx) in the inner side, I'm also using this function, but made it aware of `MajorVersion`, `MinorVersion`, `BuildNumber`, these three fields fit in with our task.

Reply:
To make sure it works, I have tested here in several ways:  
1. change `BuildNumber` in the array.
2. change `info.dwMajorVersion` and `info.dwMinorVersion`.
3. operate `setsockopt()`, `getsockopt()` under Compatibility Mode.
4. upset the order of macro `VER_SET_CONDITION()`.